### PR TITLE
migration script to populate jobStatesMap for builds and runs

### PR DESCRIPTION
**Pull request by deepikasl**
https://github.com/Shippable/base/issues/1055,
https://github.com/Shippable/base/issues/1056

### CODE REVIEW

File: migrations/pre_install/master-pre_install.sql [hunk 1]
```diff
@@ -3,3 +3,117 @@ do $$
...
+)
+SELECT "id", "projectId" as pid,"branchName" as bn,"runNumber","createdAt","statusCode","isGitTag","isPullRequest","isRelease","gitTagName","releaseName","subscriptionId", "createdBy", "updatedBy", "updatedAt"
+FROM cte
+WHERE rn = 1 and "statusCode" in (30,50,60,80) and "branchName" is not null;
+
+alter table tjsm add column "contextTypeCode" int, add column "contextValue" varchar(255), add column "lastSuccessfulJobId" varchar(24), add column "lastFailedJobId" varchar(24), add column "lastUnstableJobId" varchar(24), add column "lastTimedoutJobId" varchar(24);
+update tjsm T set "isPullRequest" = false where "isPullRequest" is null;
+update tjsm T set "releaseName" = T."gitTagName" where "isRelease" = true and "releaseName" is null;
+
+update tjsm T1 set "lastSuccessfulJobId" = T2.id from (select * from tjsm where "statusCode" = 30) T2 where
```
Comment (avinci) [inline migrations/pre_install/master-pre_install.sql]: this needs to be a join

File: migrations/pre_install/master-pre_install.sql [hunk 2]
```diff
@@ -3,3 +3,117 @@ do $$
...
+update tjsm T1 set "lastFailedJobId" = T2.id from (select * from tjsm where "statusCode" = 80) T2 where
+  ((T1."isGitTag" = T2."isGitTag") or (T1."isGitTag" is null and T2."isGitTag" is null)) and
+  ((T1."isRelease" = T2."isRelease") or (T1."isRelease" is null and T2."isRelease" is null)) and
+  ((T1."isPullRequest" = T2."isPullRequest") or (T1."isPullRequest" is null and T2."isPullRequest" is null)) and
+  ((T1."bn" = T2."bn") or (T1."bn" is null and T2."bn" is null)) and
+  ((T1."gitTagName" = T2."gitTagName") or (T1."gitTagName" is null and T2."gitTagName" is null)) and
+  ((T1."releaseName" = T2."releaseName") or (T1."releaseName" is null and T2."releaseName" is null)) and
+  T1.pid = T2.pid;
+
+delete from tjsm T1 using (select id, ROW_NUMBER() OVER (PARTITION BY "lastSuccessfulJobId", "lastUnstableJobId", "lastTimedoutJobId", "lastFailedJobId", pid order by "runNumber" desc) as rn from tjsm) T2 where T1.id = T2.id and T2.rn > 1;
```
Comment (avinci) [inline migrations/pre_install/master-pre_install.sql]: dont understand this. Why do you need to portion a delete?

File: migrations/pre_install/master-pre_install.sql (hunk 1 above)
Comment (deepikasl) [inline migrations/pre_install/master-pre_install.sql]: Here `from` acts like `join` am using self join.

File: migrations/pre_install/master-pre_install.sql (hunk 2 above)
Comment (deepikasl) [inline migrations/pre_install/master-pre_install.sql]: Am keeping only the latest records and removing the old ones.

Example:
```
 | projectId | id | statusCode | runno. | type | value | lsJobId | lfJobId | lsUJobId | lsJobId
| ---  | ---  | --- | --- | --- | --- | --- | --- | --- | --- |
| 562dbda348095b0d00ce6a43 | 562dbda348095b0d00ce1234 | 30 | 6 | gitTag | tag-v2.0 | 562dbda348095b0d00ce1234 | 562dbda348095b0d00ce1232 | 562dbda348095b0d00ce1223 | 562dbda348095b0d00ce1234

Comment (deepikasl) [inline migrations/pre_install/master-pre_install.sql]: Am keeping only the latest records and removing the old ones.

Example:

| projectId | id | statusCode | runno. | type | value | lsJobId | lfJobId | lsUJobId | lsJobId |
| ---  | ---  | --- | --- | --- | --- | --- | --- | --- | --- |
| 562dbda348095b0d00ce6a43 | 562dbda348095b0d00ce1234 | 30 | 6 | gitTag | tag-v2.0 | 562dbda348095b0d00ce1234 | 562dbda348095b0d00ce1232 | 562dbda348095b0d00ce1223 | 562dbda348095b0d00ce1234 |
| 562dbda348095b0d00ce6a43 | 562dbda348095b0d00ce1232 | 50 | 4 | gitTag | tag-v2.0 | 562dbda348095b0d00ce1234 | 562dbda348095b0d00ce1232 | 562dbda348095b0d00ce1223 | 562dbda348095b0d00ce1234 |
| 562dbda348095b0d00ce6a43 | 562dbda348095b0d00ce1223 | 60 | 4 | gitTag | tag-v2.0 | 562dbda348095b0d00ce1234 | 562dbda348095b0d00ce1232 | 562dbda348095b0d00ce1223 | 562dbda348095b0d00ce1234 |

Here according to the unique criteria of `jobStatesMap` : projectId, contextTypeCode, and contextvalue values are same but the status codes are different.

So, at first am updating the `lastSuccesfulJobId`, `lastFailedJobId`, `lastUnstableJobId`, `lastTimeOutJobId` for every record of same `projectId, contextTypeCode, and contextvalue`.
And then deleting the old or previous recordes keeping the latest one.

so after deleting the above table will have one record like this:

| projectId | id | statusCode | runno. | type | value | lsJobId | lfJobId | lsUJobId | lsJobId |
| ---  | ---  | --- | --- | --- | --- | --- | --- | --- | --- |
| 562dbda348095b0d00ce6a43 | 562dbda348095b0d00ce1234 | 30 | 6 | gitTag | tag-v2.0 | 562dbda348095b0d00ce1234 | 562dbda348095b0d00ce1232 | 562dbda348095b0d00ce1223 | 562dbda348095b0d00ce1234 |

One unique record containing all the right values.

File: migrations/post_install/master-post_install.sql [hunk 3]
```diff
@@ -3,3 +3,117 @@ create temp table temp_pa as select id from projects;
 delete from temp_pa using "projectAccounts" where "projectAccounts"."projectId" = temp_pa.id;
 update projects set "isOrphaned" = true from temp_pa where temp_pa.id = projects.id;
 drop table temp_pa;
+
+-- Migration Script to populate jobstatesMap for runs
+create temp table tjsm as WITH cte AS
```
Comment (avinci) [inline migrations/post_install/master-post_install.sql]: how do you prevent it from running multiple times?

File: migrations/post_install/master-post_install.sql [hunk 4]
```diff
@@ -3,3 +3,117 @@ create temp table temp_pa as select id from projects;
...
+(
+   SELECT "id","projectId","branchName","runNumber","createdAt","statusCode","isGitTag","isPullRequest","isRelease","gitTagName","releaseName","subscriptionId", "createdBy", "updatedBy", "updatedAt", ROW_NUMBER() OVER (PARTITION BY "projectId","branchName","statusCode","isGitTag","isPullRequest","isRelease" ORDER BY "createdAt" desc)
+AS rn FROM runs WHERE "statusCode" in (30,50,60,80) and "branchName" is not null
```
Comment (avinci) [inline migrations/post_install/master-post_install.sql]: this doesnt work by putting filters inside?

File: migrations/post_install/master-post_install.sql [hunk 5]
```diff
@@ -3,3 +3,117 @@ create temp table temp_pa as select id from projects;
...
+
+-- update contextTypeCode, contextValue
+update tjsm set "contextTypeCode" = 301, "contextValue" = tjsm."gitTagName" where "isGitTag" = true;
+update tjsm set "contextTypeCode" = 302, "contextValue" = tjsm."releaseName" where "isRelease" = true;
+update tjsm set "contextTypeCode" = 303, "contextValue" = tjsm.bn where "isPullRequest" = true;
+update tjsm set "contextTypeCode" = 304, "contextValue" = tjsm.bn where "isPullRequest" = false and "isGitTag" = false and "isRelease" = false;
+update tjsm set "updatedBy" = tjsm."createdBy" where "updatedBy" is null;
+
+-- update lastSuccessfulJobId, lastUnstableJobId, lastTimedoutJobId, lastFailedJobId
+update tjsm T1 set "lastSuccessfulJobId" = T2.id from tjsm T2 where T1."contextTypeCode" = T2."contextTypeCode" and T1."contextValue" = T2."contextValue" and T1.pid = T2.pid and T2."statusCode" = 30;
```
Comment (avinci) [inline migrations/post_install/master-post_install.sql]: still not the right query.....I am still not sure why you guys are so hesitant to write joins

UPDATE T1 
SET T1."lastSuccessfulJobId" = T2.id 
FROM tjsm T1 
JOIN tjsm T2 
ON T1."contextTypeCode" = T2."contextTypeCode" 
AND T1."contextValue" = T2."contextValue" 
AND T1.pid = T2.pid 
WHERE T2."statusCode" = 30;

File: migrations/post_install/master-post_install.sql [hunk 6]
```diff
@@ -3,3 +3,117 @@ create temp table temp_pa as select id from projects;
...
+create temp table temptjsm as WITH cte AS
+(
+   SELECT id,pid,"createdAt","contextTypeCode","contextValue","statusCode","subscriptionId", "createdBy", "updatedBy", "updatedAt","lastSuccessfulJobId","lastUnstableJobId","lastTimedoutJobId","lastFailedJobId", ROW_NUMBER() OVER (PARTITION BY "pid","contextTypeCode","contextValue" ORDER BY "createdAt" desc)
+AS rn FROM tjsm
+)
+SELECT "id",pid,"createdAt","statusCode","contextTypeCode","contextValue","lastSuccessfulJobId","lastUnstableJobId","lastTimedoutJobId","lastFailedJobId","subscriptionId", "createdBy", "updatedBy", "updatedAt"
+FROM cte
+WHERE rn = 1;
+
+truncate "jobStatesMap";
```
Comment (avinci) [inline migrations/post_install/master-post_install.sql]: if you do this, everytime your entire table will become empty for no reason.

File: migrations/post_install/master-post_install.sql [hunk 7]
```diff
@@ -3,3 +3,117 @@ create temp table temp_pa as select id from projects;
...
+
+-- add required columns to temp table tjsm
+alter table tjsm add column "contextTypeCode" int, add column "contextValue" varchar(255), add column "lastSuccessfulJobId" varchar(24), add column "lastFailedJobId" varchar(24), add column "lastUnstableJobId" varchar(24), add column "lastTimedoutJobId" varchar(24);
+
+-- update null fields
+update tjsm T set "isPullRequest" = false where "isPullRequest" is null;
+update tjsm T set "isGitTag" = false where "isGitTag" is null;
+update tjsm T set "isRelease" = false where "isRelease" is null;
+update tjsm T set "releaseName" = T."gitTagName" where "isRelease" = true and "releaseName" is null;
+
```
Comment (avinci) [inline migrations/post_install/master-post_install.sql]: if you are going to do the section below, the above section is pretty useless

File: migrations/post_install/master-post_install.sql [hunk 8]
```diff
@@ -3,3 +3,117 @@ create temp table temp_pa as select id from projects;
...
+-- update contextTypeCode, contextValue
+update rtjsm set "contextTypeCode" = 305 where "contextTypeCode" = 2010;
+update rtjsm set "contextTypeCode" = 306 where "contextTypeCode" = 2000;
+update rtjsm set "contextTypeCode" = 307 where "contextTypeCode" = 2005;
+update rtjsm set "contextTypeCode" = 308 where "contextTypeCode" = 2007;
+update rtjsm set "contextTypeCode" = 309 where "contextTypeCode" = 2012;
+update rtjsm set "contextTypeCode" = 310 where "contextTypeCode" = 2011;
+update rtjsm set "contextTypeCode" = 311 where "contextTypeCode" = 2009;
+
+update tjsm T1 set "contextValue" = T2."contextValue" from rtjsm T2 where T1.rid = T2.id;
```
Comment (avinci) [inline migrations/post_install/master-post_install.sql]: same JOIN issue

File: migrations/post_install/master-post_install.sql [hunk 9]
```diff
@@ -3,3 +3,117 @@ create temp table temp_pa as select id from projects;
...
+-- creating a temp table temptjsm which have only latest records
+create temp table temptjsm as WITH cte AS
+(
+   SELECT id,pid,"createdAt","contextTypeCode","contextValue","statusCode","subscriptionId", "createdBy", "updatedBy", "updatedAt","lastSuccessfulJobId","lastFailedJobId", ROW_NUMBER() OVER (PARTITION BY "pid","contextTypeCode","contextValue" ORDER BY "createdAt" desc)
+AS rn FROM tjsm
+)
+SELECT "id",pid,"createdAt","statusCode","contextTypeCode","contextValue","lastSuccessfulJobId","lastFailedJobId","subscriptionId", "createdBy", "updatedBy", "updatedAt"
+FROM cte
+WHERE rn = 1;
+
```
Comment (avinci) [inline migrations/post_install/master-post_install.sql]: how do you ensure that the insert will not insert a duplicate. During the time this is running, API might have processed a new job and already inserted a record into the jobStatusMap.

File: migrations/post_install/master-post_install.sql (hunk 4 above)
Comment (deepikasl) [inline migrations/post_install/master-post_install.sql]: It works.

File: migrations/post_install/master-post_install.sql (hunk 7 above)
Comment (deepikasl) [inline migrations/post_install/master-post_install.sql]: The above section is used for:  
```
update tjsm set "contextTypeCode" = 304, "contextValue" = tjsm.bn where "isPullRequest" = false and "isGitTag" = false and "isRelease" = false;
```

File: migrations/post_install/master-post_install.sql (hunk 5 above)
Comment (deepikasl) [inline migrations/post_install/master-post_install.sql]: While updating we need to specify the table name else it complains `ERROR:  relation "t1" does not exist`. 
In postgres SQL, for `update`, `join` works differently than mySql. 
In my query: from and where combination acts like join. 

If we do the above query in postgres:
```
shipdb=# UPDATE T1  SET T1."lastSuccessfulJobId" = T2.id FROM tjsm T1 JOIN tjsm T2 ON T1."contextTypeCode" = T2."contextTypeCode" AND T1."contextValue" = T2."contextValue" AND T1.pid = T2.pid WHERE T2."statusCode" = 30;
ERROR:  relation "t1" does not exist
LINE 1: UPDATE T1  SET T1."lastSuccessfulJobId" = T2.id FROM tjsm T1...
               ^
```
but this works in mySql.

Reference: https://makandracards.com/makandra/32357-postgresql-vs-mysql-how-to-update-using-a-join

File: migrations/post_install/master-post_install.sql (hunk 6 above)
Comment (deepikasl) [inline migrations/post_install/master-post_install.sql]: I run this script only once. and I truncate `jobStatesMap` only once. As `runs` already have the complete information.
My migration script takes care of it.

File: migrations/post_install/master-post_install.sql (hunk 5 above)
Comment (avinci) [inline migrations/post_install/master-post_install.sql]: Just say UPDATE

File: migrations/post_install/master-post_install.sql (hunk 6 above)
Comment (avinci) [inline migrations/post_install/master-post_install.sql]: how do you control that? Also how are you sure jobStatusMap is empty? You need to be aware that API is already up by this time and processing jobs and builds

File: migrations/post_install/master-post_install.sql (hunk 7 above)
Comment (avinci) [inline migrations/post_install/master-post_install.sql]: ok

File: migrations/post_install/master-post_install.sql (hunk 4 above)
Comment (avinci) [inline migrations/post_install/master-post_install.sql]: then why are we not doing it?

Comment (deepikasl) [inline migrations/post_install/master-post_install.sql]: Am sorry, it doesn't work that way. 

```
shipdb=# create temp table tjsm as WITH cte AS
shipdb-# (
shipdb(#    SELECT "id","projectId","branchName","runNumber","createdAt","statusCode","isGitTag","isPullRequest","isRelease","gitTagName","releaseName","subscriptionId", "createdBy", "updatedBy", "updatedAt", ROW_NUMBER() OVER (PARTITION BY "projectId","branchName","statusCode","isGitTag","isPullRequest","isRelease" WHERE "statusCode" in (30,50,60,80) and "branchName" is not null ORDER BY "createdAt" desc)
shipdb(# AS rn FROM runs
shipdb(# )
shipdb-# SELECT "id", "projectId" as pid,"branchName" as bn,"runNumber","createdAt","statusCode","isGitTag","isPullRequest","isRelease","gitTagName","releaseName","subscriptionId", "createdBy", "updatedBy", "updatedAt"
shipdb-# FROM cte
shipdb-# WHERE rn = 1;
ERROR:  syntax error at or near "WHERE"
LINE 3: ...tatusCode","isGitTag","isPullRequest","isRelease" WHERE "sta...
                                                             ^
```

Comment (avinci) [inline migrations/post_install/master-post_install.sql]: ok then its fine.

File: migrations/post_install/master-post_install.sql (hunk 9 above)
Comment (deepikasl) [inline migrations/post_install/master-post_install.sql]: I have updated it to `insert if not present else update`

File: migrations/post_install/master-post_install.sql (hunk 5 above)
Comment (avinci) [inline migrations/post_install/master-post_install.sql]: Did we try it without adding the alias?

File: migrations/post_install/master-post_install.sql (hunk 3 above)
Comment (avinci) [inline migrations/post_install/master-post_install.sql]: this is still not addressed. How do you prevent this query from running again if the installer is run again

File: migrations/post_install/master-post_install.sql (hunk 5 above)
Comment (deepikasl) [inline migrations/post_install/master-post_install.sql]: I have tested for these three queries. But it doesn't work for any one of them.
```shipdb=# 
shipdb=# UPDATE T1 SET T1."lastSuccessfulJobId" = T2.id FROM tjsm T1 JOIN tjsm T2 ON T1."contextTypeCode" = T2."contextTypeCode" AND T1."contextValue" = T2."contextValue" AND T1.pid = T2.pid WHERE T2."statusCode" = 30;
ERROR:  relation "t1" does not exist
LINE 1: UPDATE T1 SET T1."lastSuccessfulJobId" = T2.id FROM tjsm T1 ...
               ^
```

With just `update`:
```
shipdb=# UPDATE SET T1."lastSuccessfulJobId" = T2.id FROM tjsm T1 JOIN tjsm T2 ON T1."contextTypeCode" = T2."contextTypeCode" AND T1."contextValue" = T2."contextValue" AND T1.pid = T2.pid WHERE T2."statusCode" = 30;
ERROR:  syntax error at or near "."
LINE 1: UPDATE SET T1."lastSuccessfulJobId" = T2.id FROM tjsm T1 JOI...
                     ^
```

```
shipdb=# UPDATE T1 FROM tjsm T1 JOIN tjsm T2 ON T1."contextTypeCode" = T2."contextTypeCode" AND T1."contextValue" = T2."contextValue" AND T1.pid = T2.pid SET T1."lastSuccessfulJobId" = T2.id WHERE T2."statusCode" = 30;
ERROR:  syntax error at or near "FROM"
LINE 1: UPDATE T1 FROM tjsm T1 JOIN tjsm T2 ON T1."contextTypeCode" ...
                  ^
```

File: migrations/post_install/master-post_install.sql (hunk 3 above)
Comment (deepikasl) [inline migrations/post_install/master-post_install.sql]: I have a thought: 
```
do $$
 begin
    if not exists (select 1 from runs T1, "jobStatesMap" T2 where T1."createdAt" = T2."createdAt" order by T1."createdAt" asc limit 1)
    -- then do migartion
    end
  end 
$$;

```
But this uses `do $$ begin`.

--- 
Else, we can create a new file and put this script in it and run it only once after installation. 
And specify it in release notes that this script needs to be run after installation.

Comment (avinci) [inline migrations/post_install/master-post_install.sql]: This script is very nasty. Can you tell me why? Second, dont understand why do we need a do?

This is the way you can do it. 
https://www.postgresql.org/docs/9.1/static/datatype-datetime.html

IF (SELECT 1 FROM jobStatusMap 
WHERE convert this to midnight of the date(MIN("createdAt") AS "createdAt")  = midnight of date( GETDATE())")
THEN

Comment (deepikasl) [inline migrations/post_install/master-post_install.sql]: There are no procedural elements in standard SQL. The `IF` statement is part of the default procedural language pgSQL. we need to create a function or execute an ad-hoc statement with the `DO` command.

**Without do**:
```
shipdb=# if not exists (select 1 from runs T1, "jobStatesMap" T2 where T1."createdAt" = T2."createdAt" order by T1."createdAt" asc limit 1) then
shipdb-#   create temp table tjsm as WITH cte AS
shipdb-#   (
shipdb(#      SELECT "id","projectId","branchName","runNumber","createdAt","statusCode","isGitTag","isPullRequest","isRelease","gitTagName","releaseName","subscriptionId", "createdBy", "updatedBy", "updatedAt", ROW_NUMBER() OVER (PARTITION BY "projectId","branchName","statusCode","isGitTag","isPullRequest","isRelease" ORDER BY "createdAt" desc)
shipdb(#   AS rn FROM runs WHERE "statusCode" in (30,50,60,80) and "branchName" is not null
shipdb(#   )
shipdb-#   SELECT "id", "projectId" as pid,"branchName" as bn,"runNumber","createdAt","statusCode","isGitTag","isPullRequest","isRelease","gitTagName","releaseName","subscriptionId", "createdBy", "updatedBy", "updatedAt"
shipdb-#   FROM cte
shipdb-#   WHERE rn = 1;
ERROR:  syntax error at or near "if"
LINE 1: if not exists (select 1 from runs T1, "jobStatesMap" T2 wher...
        ^
shipdb=# end if;
ERROR:  syntax error at or near "if"
LINE 1: end if;
            ^
```
---
**With do:**
```
shipdb=# do $$
shipdb$#   begin
shipdb$#     if not exists(select 1 from runs T1, "jobStatesMap" T2 where T1."createdAt" = T2."createdAt" order by T1."createdAt" asc limit 1) then
shipdb$#       create temp table tjsm as WITH cte AS
shipdb$#       (
shipdb$#          SELECT "id","projectId","branchName","runNumber","createdAt","statusCode","isGitTag","isPullRequest","isRelease","gitTagName","releaseName","subscriptionId", "createdBy", "updatedBy", "updatedAt", ROW_NUMBER() OVER (PARTITION BY "projectId","branchName","statusCode","isGitTag","isPullRequest","isRelease" ORDER BY "createdAt" desc)
shipdb$#       AS rn FROM runs WHERE "statusCode" in (30,50,60,80) and "branchName" is not null
shipdb$#       )
shipdb$#       SELECT "id", "projectId" as pid,"branchName" as bn,"runNumber","createdAt","statusCode","isGitTag","isPullRequest","isRelease","gitTagName","releaseName","subscriptionId", "createdBy", "updatedBy", "updatedAt"
shipdb$#       FROM cte
shipdb$#       WHERE rn = 1;
shipdb$#     end if;
shipdb$#   end
shipdb$# $$;
DO
```

File: migrations/post_install/master-post_install.sql [hunk 10]
```diff
@@ -3,3 +3,137 @@ create temp table temp_pa as select id from projects;
...
+      update "jobStatesMap" T1 set "lastSuccessfulJobId" = T2."lastSuccessfulJobId" from temptjsm T2 where T1."projectId" = T2.pid and T1."contextTypeCode" = T2."contextTypeCode" and T1."contextValue" = T2."contextValue" and T1."lastSuccessfulJobId" is null and T2."lastSuccessfulJobId" is not null;
+
+      update "jobStatesMap" T1 set "lastFailedJobId" = T2."lastFailedJobId" from temptjsm T2 where T1."projectId" = T2.pid and T1."contextTypeCode" = T2."contextTypeCode" and T1."contextValue" = T2."contextValue" and T1."lastFailedJobId" is null and T2."lastFailedJobId" is not null;
+
+      drop table tjsm;
+      drop table rtjsm;
+      drop table temptjsm;
+    end if;
+  end
+$$;
```
Comment (deepikasl) [inline migrations/post_install/master-post_install.sql]: It took `Time: 92.105 ms` to run complete script.
And populated `571` records of "jobStatesMap" for `4097` records of runs and `1994` records of builds.

**Pull request merged**